### ISSUE TITLE
[IOS-XE] Parser for "show mka sessions" does not work

### ISSUE THREAD

**Issue by ChristopherJHart** [CiscoTestAutomation/genieparser#851]
Sample CLI output:
```
R1#show mka sessions 

Total MKA Sessions....... 2
      Secured Sessions... 2
      Pending Sessions... 0

====================================================================================================
Interface       Local-TxSCI          Policy-Name       Inherited          Key-Server                                            
Port-ID         Peer-RxSCI           MACsec-Peers      Status             CKN                                                   
====================================================================================================
Te0/1/0         8c94.61ea.0e08/001d  MACSEC_POLICY_1   NO                 NO                                                    
29              04bd.97d1.2e20/0082  1                 Secured            1A                                                              

Te0/1/3         8c94.61ea.0e0b/001f  MACSEC_POLICY_1   NO                 NO                                                    
31              04bd.97d1.2d69/0082  1                 Secured            1A   

```
Traceback:
```
2024-04-29T13:12:37: %AETEST-ERROR: Caught an exception while executing section ShowMkaSessions:
2024-04-29T13:12:37: %AETEST-ERROR: Traceback (most recent call last):
2024-04-29T13:12:37: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/unittests.py", line 499, in test
2024-04-29T13:12:37: %AETEST-ERROR:     self.test_golden(golden_steps, local_class,
2024-04-29T13:12:37: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/unittests.py", line 461, in wrapper
2024-04-29T13:12:37: %AETEST-ERROR:     func(self, *args, **kwargs)
2024-04-29T13:12:37: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/unittests.py", line 604, in test_golden
2024-04-29T13:12:37: %AETEST-ERROR:     parsed_json_data = format_output(parsed_output)
2024-04-29T13:12:37: %AETEST-ERROR:                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-04-29T13:12:37: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/common.py", line 239, in format_output
2024-04-29T13:12:37: %AETEST-ERROR:     v = format_output(v, tab + 2) if isinstance(v, dict) else repr(v)
2024-04-29T13:12:37: %AETEST-ERROR:         ^^^^^^^^^^^^^^^^^^^^^^^^^
2024-04-29T13:12:37: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/common.py", line 238, in format_output
2024-04-29T13:12:37: %AETEST-ERROR:     for k, v in sorted(parser_data.items()):
2024-04-29T13:12:37: %AETEST-ERROR:                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-04-29T13:12:37: %AETEST-ERROR: TypeError: '<' not supported between instances of 'int' and 'str'
```

**Comment by SohanTirpude** [comment]
Hello @ChristopherJHart,

It looks like a bug. Kindly give me some time to analyse this issue.

Thank you.

**Comment by SohanTirpude** [comment]
Hello @ChristopherJHart,

Well, I am not seeing any issue with this parser. This is the parser output I am getting:

```
{
    "sessions": {
        "total-mka-sessions": 2,
        "secured-mka-sessions": 2,
        "pending-mka-sessions": 0,
        1: {
            "interface": "Te0/1/0",
            "local-txsci": "8c94.61ea.0e08/001d",
            "policy-name": "MACSEC_POLICY_1",
            "inherited": "NO",
            "key-server": "NO",
            "port-id": "29",
            "peer-rxsci": "04bd.97d1.2e20/0082",
            "macsec-peers": "1",
            "status": "Secured",
            "ckn": "1A"
        },
        2: {
            "interface": "Te0/1/3",
            "local-txsci": "8c94.61ea.0e0b/001f",
            "policy-name": "MACSEC_POLICY_1",
            "inherited": "NO",
            "key-server": "NO",
            "port-id": "31",
            "peer-rxsci": "04bd.97d1.2d69/0082",
            "macsec-peers": "1",
            "status": "Secured",
            "ckn": "1A"
        }
    }
}
```

Kindly re-check again and let me know.

Thank you.

**Comment by SohanTirpude** [comment]
Hello @ChristopherJHart,

I am not seeing any issue parsing this output. So, kindly recheck it.

Thank you.

**Comment by SohanTirpude** [comment]
Hello @ChristopherJHart,

You are not responding; hence I am closing this ticket. I hope you are not facing this issue anymore, but if you are still facing the same issue, kindly let me know, I'll re-open the ticket.

If you are facing any other issues, kindly open a new issue.

Thank you.

**Comment by ChristopherJHart** [comment]
@SohanTirpude Odd - if I create a unit test with this exact output, I see it fail.

```
(venv) ➜  tests git:(dev) ✗ python3 --version
Python 3.11.9

(venv) ➜  tests git:(dev) ✗ pwd
/Users/chart2/Documents/Python/genieparser/tests

(venv) ➜  tests git:(dev) ✗ cat iosxe/ShowMkaSessions/cli/equal/golden_output_3_output.txt
R1#show mka sessions

Total MKA Sessions....... 2
      Secured Sessions... 2
      Pending Sessions... 0

====================================================================================================
Interface       Local-TxSCI          Policy-Name       Inherited          Key-Server
Port-ID         Peer-RxSCI           MACsec-Peers      Status             CKN
====================================================================================================
Te0/1/0         8c94.61ea.0e08/001d  MACSEC_POLICY_1   NO                 NO
29              04bd.97d1.2e20/0082  1                 Secured            1A

Te0/1/3         8c94.61ea.0e0b/001f  MACSEC_POLICY_1   NO                 NO
31              04bd.97d1.2d69/0082  1                 Secured            1A

(venv) ➜  tests git:(dev) ✗ cat iosxe/ShowMkaSessions/cli/equal/golden_output_3_expected.py
expected_output = {}

(venv) ➜  tests git:(dev) ✗ python folder_parsing_job.py -o iosxe -c ShowMkaSessions
<module 'pyats.aetest.testscript.unittests' from '/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/unittests.py'>
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: |                   Starting testcase SuperFileBasedTesting                    |
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: |                            Starting section setup                            |
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: The result of section setup is => PASSED
2024-05-10T12:56:11: %AETEST-INFO: The result of testcase SuperFileBasedTesting is => PASSED
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: |                           Starting testcase iosxe                            |
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: |                            Starting section setup                            |
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: The result of section setup is => PASSED
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: |                       Starting section ShowMkaSessions                       |
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: +..............................................................................+
2024-05-10T12:56:11: %AETEST-INFO: :                  Starting STEP 1: iosxe -> ShowMkaSessions                   :
2024-05-10T12:56:11: %AETEST-INFO: +..............................................................................+
2024-05-10T12:56:11: %AETEST-INFO: +..............................................................................+
2024-05-10T12:56:11: %AETEST-INFO: :          Starting STEP 1.1: Test Golden -> iosxe -> ShowMkaSessions          :
2024-05-10T12:56:11: %AETEST-INFO: +..............................................................................+
2024-05-10T12:56:11: %AETEST-INFO: +..............................................................................+
2024-05-10T12:56:11: %AETEST-INFO: :    Starting STEP 1.1.1: Gold -> iosxe -> ShowMkaSessions -> golden_output    :
2024-05-10T12:56:11: %AETEST-INFO: :                                      _2                                      :
2024-05-10T12:56:11: %AETEST-INFO: +..............................................................................+
2024-05-10T12:56:11: %AETEST-INFO: The result of STEP 1.1.1: Gold -> iosxe -> ShowMkaSessions -> golden_output_2 is => PASSED
2024-05-10T12:56:11: %AETEST-INFO: +..............................................................................+
2024-05-10T12:56:11: %AETEST-INFO: :    Starting STEP 1.1.2: Gold -> iosxe -> ShowMkaSessions -> golden_output    :
2024-05-10T12:56:11: %AETEST-INFO: :                                      _3                                      :
2024-05-10T12:56:11: %AETEST-INFO: +..............................................................................+
2024-05-10T12:56:11: %AETEST-INFO: The result of STEP 1.1.2: Gold -> iosxe -> ShowMkaSessions -> golden_output_3 is => ERRORED
2024-05-10T12:56:11: %AETEST-INFO: The result of STEP 1.1: Test Golden -> iosxe -> ShowMkaSessions is => ERRORED
2024-05-10T12:56:11: %AETEST-INFO: The result of STEP 1: iosxe -> ShowMkaSessions is => ERRORED
2024-05-10T12:56:11: %AETEST-ERROR: Caught an exception while executing section ShowMkaSessions:
2024-05-10T12:56:11: %AETEST-ERROR: Traceback (most recent call last):
2024-05-10T12:56:11: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/unittests.py", line 499, in test
2024-05-10T12:56:11: %AETEST-ERROR:     self.test_golden(golden_steps, local_class,
2024-05-10T12:56:11: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/unittests.py", line 461, in wrapper
2024-05-10T12:56:11: %AETEST-ERROR:     func(self, *args, **kwargs)
2024-05-10T12:56:11: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/unittests.py", line 604, in test_golden
2024-05-10T12:56:11: %AETEST-ERROR:     parsed_json_data = format_output(parsed_output)
2024-05-10T12:56:11: %AETEST-ERROR:                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-05-10T12:56:11: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/common.py", line 191, in format_output
2024-05-10T12:56:11: %AETEST-ERROR:     v = format_output(v, tab + 2) if isinstance(v, dict) else repr(v)
2024-05-10T12:56:11: %AETEST-ERROR:         ^^^^^^^^^^^^^^^^^^^^^^^^^
2024-05-10T12:56:11: %AETEST-ERROR:   File "/Users/chart2/Documents/Python/genieparser/src/genie/libs/parser/utils/common.py", line 190, in format_output
2024-05-10T12:56:11: %AETEST-ERROR:     for k, v in sorted(parser_data.items()):
2024-05-10T12:56:11: %AETEST-ERROR:                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-05-10T12:56:11: %AETEST-ERROR: TypeError: '<' not supported between instances of 'int' and 'str'
2024-05-10T12:56:11: %AETEST-INFO: +..........................................................+
2024-05-10T12:56:11: %AETEST-INFO: :                       STEPS Report                       :
2024-05-10T12:56:11: %AETEST-INFO: +..........................................................+
2024-05-10T12:56:11: %AETEST-INFO: STEP 1 - iosxe -> ShowMkaSessions                 Errored
2024-05-10T12:56:11: %AETEST-INFO: STEP 1.1 - Test Golden -> iosxe -> ShowMkaSessionsErrored
2024-05-10T12:56:11: %AETEST-INFO: STEP 1.1.1 - Gold -> iosxe -> ShowMkaSessions -> golden_output_2Passed
2024-05-10T12:56:11: %AETEST-INFO: STEP 1.1.2 - Gold -> iosxe -> ShowMkaSessions -> golden_output_3Errored
2024-05-10T12:56:11: %AETEST-INFO: ............................................................
2024-05-10T12:56:11: %AETEST-INFO: The result of section ShowMkaSessions is => ERRORED
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: |                           Starting section cleanup                           |
2024-05-10T12:56:11: %AETEST-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %AETEST-INFO: The result of section cleanup is => PASSED
2024-05-10T12:56:11: %AETEST-INFO: The result of testcase iosxe is => ERRORED
2024-05-10T12:56:11: %GENIE-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %GENIE-INFO: |                               Unittest results                               |
2024-05-10T12:56:11: %GENIE-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %GENIE-INFO:  SECTIONS/TESTCASES                                                      RESULT
2024-05-10T12:56:11: %GENIE-INFO: --------------------------------------------------------------------------------
2024-05-10T12:56:11: %GENIE-INFO: .
2024-05-10T12:56:11: %GENIE-INFO: |-- SuperFileBasedTesting                                                 PASSED
2024-05-10T12:56:11: %GENIE-INFO: |   `-- setup                                                             PASSED
2024-05-10T12:56:11: %GENIE-INFO: `-- iosxe                                                                ERRORED
2024-05-10T12:56:11: %GENIE-INFO:     |-- setup                                                             PASSED
2024-05-10T12:56:11: %GENIE-INFO:     |-- ShowMkaSessions                                                  ERRORED
2024-05-10T12:56:11: %GENIE-INFO:     |   |-- Step 1: iosxe -> ShowMkaSessions                             ERRORED
2024-05-10T12:56:11: %GENIE-INFO:     |   |-- Step 1.1: Test Golden -> iosxe -> ShowMkaSessions            ERRORED
2024-05-10T12:56:11: %GENIE-INFO:     |   |-- Step 1.1.1: Gold -> iosxe -> ShowMkaSessions -> golden_...    PASSED
2024-05-10T12:56:11: %GENIE-INFO:     |   `-- Step 1.1.2: Gold -> iosxe -> ShowMkaSessions -> golden_...   ERRORED
2024-05-10T12:56:11: %GENIE-INFO:     `-- cleanup                                                           PASSED
2024-05-10T12:56:11: %GENIE-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %GENIE-INFO: |                                   Summary                                    |
2024-05-10T12:56:11: %GENIE-INFO: +------------------------------------------------------------------------------+
2024-05-10T12:56:11: %GENIE-INFO:  Number of ABORTED                                                            0
2024-05-10T12:56:11: %GENIE-INFO:  Number of BLOCKED                                                            0
2024-05-10T12:56:11: %GENIE-INFO:  Number of ERRORED                                                            1
2024-05-10T12:56:11: %GENIE-INFO:  Number of FAILED                                                             0
2024-05-10T12:56:11: %GENIE-INFO:  Number of PASSED                                                             1
2024-05-10T12:56:11: %GENIE-INFO:  Number of PASSX                                                              0
2024-05-10T12:56:11: %GENIE-INFO:  Number of SKIPPED                                                            0
2024-05-10T12:56:11: %GENIE-INFO:  Total Number                                                                 2
2024-05-10T12:56:11: %GENIE-INFO:  Success Rate                                                             50.0%
2024-05-10T12:56:11: %GENIE-INFO: --------------------------------------------------------------------------------
2024-05-10T12:56:11: %GENIE-INFO:  Total Passing Unittests                                                      1
2024-05-10T12:56:11: %GENIE-INFO:  Total Failed Unittests                                                       1
2024-05-10T12:56:11: %GENIE-INFO:  Total Errored Unittests                                                      0
2024-05-10T12:56:11: %GENIE-INFO:  Total Unittests                                                              2
2024-05-10T12:56:11: %GENIE-INFO: --------------------------------------------------------------------------------
```

**Comment by SohanTirpude** [comment]
Hello @ChristopherJHart,

Yes, if I follow the testing the way you are doing, then I am able to reproduce this issue. But this is not a proper way to test it. Kindly use the below script to test you device output.

```
from pyats.topology.loader import load
 
testbed = """
devices:
  mock_device:
    os: 'iosxe'  # Change 'os' as per the requirement
    connections:
      defaults:
        class: unicon.Unicon
"""

testbed = load(testbed)

dev = testbed.devices['mock_device'] 

output = '''R1#show mka sessions

Total MKA Sessions....... 2
      Secured Sessions... 2
      Pending Sessions... 0

====================================================================================================
Interface       Local-TxSCI          Policy-Name       Inherited          Key-Server
Port-ID         Peer-RxSCI           MACsec-Peers      Status             CKN
====================================================================================================
Te0/1/0         8c94.61ea.0e08/001d  MACSEC_POLICY_1   NO                 NO
29              04bd.97d1.2e20/0082  1                 Secured            1A

Te0/1/3         8c94.61ea.0e0b/001f  MACSEC_POLICY_1   NO                 NO
31              04bd.97d1.2d69/0082  1                 Secured            1A
'''

output = dev.parse("show mka sessions", output=output)
print(output)
```
With this above script, I am not seeing any error. So, kindly try this approach and let me know the update.

Thank you.

**Comment by SohanTirpude** [comment]
Hello @ChristopherJHart,

Is there any update on this? Kindly let me know.

Thank you.

**Comment by ChristopherJHart** [comment]
Hi @SohanTirpude - that script does not work for me:
```
(venv) ➜  genieparser git:(dev) ✗ cat test.py
from pyats.topology.loader import load

testbed = """
devices:
  mock_device:
    os: 'iosxe'
    connections:
      defaults:
        class: unicon.Unicon
"""

testbed = load(testbed)

dev = testbed.devices["mock_device"]

output = """R1#show mka sessions

Total MKA Sessions....... 2
      Secured Sessions... 2
      Pending Sessions... 0

====================================================================================================
Interface       Local-TxSCI          Policy-Name       Inherited          Key-Server
Port-ID         Peer-RxSCI           MACsec-Peers      Status             CKN
====================================================================================================
Te0/1/0         8c94.61ea.0e08/001d  MACSEC_POLICY_1   NO                 NO
29              04bd.97d1.2e20/0082  1                 Secured            1A

Te0/1/3         8c94.61ea.0e0b/001f  MACSEC_POLICY_1   NO                 NO
31              04bd.97d1.2d69/0082  1                 Secured            1A
"""

output = dev.parse("show mka sessions", output=output)
print(output)
(venv) ➜  genieparser git:(dev) ✗ python3 test.py  
Could not convert testbed, Extended Device functionality will not be available: Argument 'key' has incorrect type (expected str, got NoneType)
Traceback (most recent call last):
  File "src/pyats/topology/loader/base.py", line 115, in pyats.topology.loader.base.BaseTestbedLoader.create_testbed
  File "src/genie/testbed.py", line 4, in genie.testbed.load
  File "src/genie/conf/main.py", line 97, in genie.conf.main.Genie.init
  File "src/genie/conf/utils/converter.py", line 77, in genie.conf.utils.converter.Converter.convert_tb
  File "src/genie/conf/utils/converter.py", line 168, in genie.conf.utils.converter.Converter.convert_device
  File "src/pyats/topology/factory.py", line 73, in pyats.topology.factory.TopologyMetaClassFactory.__call__
  File "src/pyats/datastructures/factory.py", line 56, in pyats.datastructures.factory.MetaClassFactory.__call__
  File "/Users/chart2/Documents/Python/genieparser/venv/lib/python3.11/site-packages/genie/libs/conf/device/iosxe/device.py", line 101, in __init__
    super().__init__(*args, **kwargs)
  File "/Users/chart2/Documents/Python/genieparser/venv/lib/python3.11/site-packages/genie/libs/conf/device/cisco/device.py", line 28, in __init__
    super().__init__(*args, **kwargs)
  File "/Users/chart2/Documents/Python/genieparser/venv/lib/python3.11/site-packages/genie/libs/conf/device/__init__.py", line 230, in __init__
    super().__init__(*args, **kwargs)
  File "src/genie/conf/base/device.py", line 121, in genie.conf.base.device.Device.__init__
  File "src/genie/conf/base/api.py", line 163, in genie.conf.base.api.API.__init__
  File "src/genie/conf/base/api.py", line 102, in genie.conf.base.api._load_function_json
  File "src/genie/abstract/package.py", line 111, in genie.abstract.package.AbstractTree.from_json
  File "src/genie/abstract/package.py", line 179, in genie.abstract.package.AbstractTreeNode._from_json
TypeError: Argument 'key' has incorrect type (expected str, got NoneType)
Traceback (most recent call last):
  File "/Users/chart2/Documents/Python/genieparser/test.py", line 33, in <module>
    output = dev.parse("show mka sessions", output=output)
             ^^^^^^^^^
  File "src/pyats/topology/device.py", line 565, in pyats.topology.device.Device.__getattr__
AttributeError: 'Device' object has no attribute 'parse'
```
Also, you stated:

> Yes, if I follow the testing the way you are doing, then I am able to reproduce this issue. But this is not a proper way to test it.

If unit tests for a specific parser that test how a parser parses specific output from a device are not the proper way to test a parser, then respectfully, why do the unit tests exist? 😅

**Comment by SohanTirpude** [comment]
Hello @ChristopherJHart,

Yes, you were right. We have found the cause of this issue and raised the PR for it. Once it is merged, I will update you back.

Thank you.

**Comment by SohanTirpude** [comment]
Hello @ChristopherJHart,

The PR has been merged and the fix will be available in the upcoming 24.6 release. Hence, I am closing this issue for now. If you are still seeing this issue with 24.6 then kindly re-open the issue.

Thank you.
